### ISSUE TITLE
Ajax page reload

### ISSUE THREAD

**Issue by Brunowilliang** [elementor/elementor#11493]
loading pages by ajax

**Comment by marekka** [comment]
+1

**Comment by willryanuk** [comment]
This would make Elementor a lot more powerful. We often write Single Page Applications SPAs, with each page loading dynamically over ajax, and right now I can't see a way of loading Elementor content lazily.

**Comment by SuperstrongBE** [comment]
I'm looking through the code to find a way to do that! It's amazing that elementor teams don't give feedback about this feature! 

If i understand the mecanism, every widget carry his own **data-widget_type** and the value of this field is the {elementType.skinName} used in the hook: frontend/element_ready/{elementType.skinName}.

These hook is triggered somewhere but i can't find how to retrigger from the elementorFrontend available form the 
```window.elementorFrontend```

I've tested some basic code and it works
```jQuery( window ).on( 'elementor/frontend/init', () => {

    console.log ('front end init',elementorFrontend);

    elementorFrontend.hooks.addAction( 'frontend/element_ready/image-carousel.default', (instance)=>{

        instance.documents = {}
        console.log("that kind of hook work",instance)

    });
    
});```
But no clue on how to trigger these frontend/element_ready/{elementType.skinName}.

**Comment by dcooney** [comment]
@willryanuk I've spent a lot of time trying to get Elementor working with my [Ajax Load More](https://connekthq.com/plugins/ajax-load-more/) plugin. 

I found that simply triggering `window.elementorFrontend.init();` will reload the entire frontend and start up all elements. ( TOC, Sliders, Main Nav menus etc).

This works if you are replacing the entire page of content, but if you are looking to trigger it on certain elements this method does not work and it initiates widgets multiple times.

A few other important functions I found were, `elementorFrontend.initOnReadyElements()` and `elementorFrontend.initOnReadyComponents()`. Both of these can be run outside of the `init()` function.

@SuperstrongBE What in your code above works? The `console.log`? It doesn't start the element though?

I've combed thru a couple of widgets in depth, the Table of Contents and Social Share. Both widgets are mess of unnecessary JavaScript and they both provide no way to filter the selectors so they are always scoped to the global `window` object.

**Comment by SuperstrongBE** [comment]
@dcooney thank for your feedback. The first line was missing in my code.
@maxjutras from FB elementor community kindly pointed me on this issue and he mention a qucik guide about implementing elementor with barba.js (with a bonus for sage users). https://github.com/elementor/elementor/issues/5843?fbclid=IwAR2gHzFXEhPuyE0xt9eiOnBm6ZDRtRVNkBi7kjJcejS7oSl08OdXoYoSjec

**Comment by SuperstrongBE** [comment]
Oh Gosh it was simple as that https://github.com/elementor/elementor/issues/5843 !

**Comment by willryanuk** [comment]
@dcooney thanks for looking into this. I've read many threads about using window.elementorFrontend.init(); to reinitialise elements... luckily for me, I am doing full page reloads (not loading more as such) so maybe this will work for my purposes, that being said, i feel i've read reports that some Elementor elements do not re-initialise properly with this technique.

It's a shame that Elementor don't see the use case of treating re-initialisation as quite a powerful feature for a lot of applications. Every time i approach this problem, it makes me think the only way to achieve what i want is by loading content between dynamically between two iFrames, but this is obviously quite a disgusting approach(!!).

**Comment by SuperstrongBE** [comment]
Yes, the elementor mechanism is not ready for Ajax uses. But the init process is easy to adapt (from the dev team perspective). 

- Widget controls in the advanced tab should have an option to be flagged as "sticky" to prevent the reinit process when called (eg: usefull for menu, popup, cart ).
- Containers (column, section) should have a "ajax target" id to allow injecting ajax content on the fly ! 
- Button, link, and other "page reload behavior" element must have an "ajax options" set to define where content behind the link must be injected. 

These feature could make elementor achieive the "next level content management for wordpress"!
To rid of these "missing behaviors", I'm used to make a child theme of hello elementor, and set my non-reloading elements (menu and others) outside the elementor's scope. But we have to keep in mind to not break basic SEO behaviors.

**Comment by willryanuk** [comment]
@SuperstrongBE Thanks for joining the discussion. When you say setting non-reloading elements (e.g. menus) to "outside of elementor's scope", how do you achieve this? From what I can see, the init function appears to be on a global scope i.e. `window.elementorFrontend.init()` so it doesn't seem like there is a good way to avoid re-initialising certain components over and over.

Also, does `window.elementorFrontend.init()` work consistently in your experience for all Elementor components? I've read about problems with popovers and accordions.

**Comment by SuperstrongBE** [comment]
@willryanuk when i say « outside the elementor scope » I mean outside the_content, so the menu is not elementor widget but a home made component on my child theme

**Comment by SuperstrongBE** [comment]
Hi, i have opened a new feature request for AJAX handling 
https://github.com/elementor/elementor/issues/13252

**Comment by ofmarconi** [comment]
+1
